### PR TITLE
feat: #1018 add auth test for bceid and bcsc login

### PR DESCRIPTION
**Pull request by MCatherine1994**
- add auth test for bceid and bcsc login
- add display name to the audit log for bcsc login, cause bcsc login doesn't return idp username

### CODE REVIEW

File: server/auth_function/readme.md [hunk 1]
```diff
@@ -35,6 +35,7 @@ sudo dockerd
 cd server/auth_function
 pytest
 ```
+Note: We need to use master db user(postgres) to run the tests.
```
Comment (MCatherine1994) [inline server/auth_function/readme.md:38]: currently the auth function needs access to fam_user and fam_role table. Later one once we read admin role, we need access to admin tables. So only master db user can have access to all those tables.

File: server/auth_function/test/conftest.py [hunk 2]
```diff
@@ -96,46 +94,8 @@ def cognito_context():
     yield context
 
 
-@pytest.fixture(scope="session")
-def test_user_properties(cognito_event):
```
Comment (MCatherine1994) [inline server/auth_function/test/conftest.py:100]: Remove these fixture methods, as they rely on other fixture methods that is parametrized. Refactor them into the util file as regular function, so we can pass parameter in.

File: server/auth_function/test/lamda_function_test.py [hunk 3]
```diff
@@ -5,20 +5,34 @@
...
+@pytest.mark.parametrize(
+    "cognito_event",
+    [
+        "login_event.json",
+        "login_event_bceid.json",
+        "login_event_bcsc.json",
+    ],
+    indirect=["cognito_event"],
+)
+def test_create_user_if_not_found(db_pg_transaction, cognito_event, cognito_context):
```
Comment (MCatherine1994) [inline server/auth_function/test/lamda_function_test.py:33]: parametrize the cognito_event fixture method, so by given different json file, it will return different data to run the test

File: server/auth_function/test/conftest.py [hunk 4]
```diff
@@ -80,8 +78,8 @@ def db_pg_transaction(db_pg_connection, monkeypatch):
 
 # Load a sample cognito test event if necessary
 @pytest.fixture(scope="session")
-def cognito_event():
-    event_file_path = os.path.join(os.path.dirname(__file__), "login_event.json")
+def cognito_event(request):
```
Comment (ianliuwk1019) [inline server/auth_function/test/conftest.py:81]: Just curious, I assume "request" isn't the http request but the pytest internal object?

File: server/auth_function/test/util.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    # For Insert statement, pass parameters as .execute()'s second arguments so they get proper sanitization.
+    cursor.execute(
+        raw_query,
+        (
+            test_user_properties.get("idp_username"),
+            test_user_properties.get("idp_type_code"),
+            TEST_ROLE_NAME,
+        ),
+    )
```
Comment (ianliuwk1019) [inline server/auth_function/test/util.py:135]: Just curious, were these functions previous from `conftest.py` not easy to use (despite you refactored a little on code)? 
All of these functions migrated into utils.py seems very much like dependencies before running tests, which is one thing pytest's "**@pytest.fixture**" is good at and convenient to use. And fixture can contain other fixtures, the function in test does not need to call all of them(the setup/dependencies).

File: server/auth_function/test/lamda_function_test.py [hunk 6]
```diff
@@ -5,20 +5,34 @@
...
-):
 
+@pytest.mark.parametrize(
+    "cognito_event",
+    [
+        "login_event.json",
+        "login_event_bceid.json",
+        "login_event_bcsc.json",
+    ],
+    indirect=["cognito_event"],
```
[the remark this answers is not in the document]
Comment (MCatherine1994) [reply server/auth_function/test/lamda_function_test.py:31]: So I have to set indirect in order to pass parameter to fixture methods. Oh but I can just set it to be true, cause we only one parameter for cognito_event here. 

Using the indirect=True parameter when parametrizing a test allows to parametrize a test with a fixture receiving the values before passing them to a test.

File: server/auth_function/test/conftest.py (hunk 4 above)
Comment (MCatherine1994) [reply server/auth_function/test/conftest.py:81]: this is how we can read the parameter passed to fixture method

File: server/auth_function/test/util.py (hunk 5 above)
Comment (MCatherine1994) [reply server/auth_function/test/util.py:135]: oh yeah, initially I need to pass some parameter into the cognito_event fixture, but you're right, let me try again, maybe it's now working.

Comment (MCatherine1994) [reply server/auth_function/test/util.py]: Hi Ian, you're right, even we pass parameters the fixture contains another fixture still works as expected, I switched back, thanks so much for noticing this!!

File: server/auth_function/lambda_function.py [hunk 7]
```diff
@@ -90,6 +89,10 @@ def lambda_handler(event: event_type.Event, context: Any) -> event_type.Event:
             audit_event_log["requestingUser"]["idpUserName"] = event["request"][
                 "userAttributes"
             ]["custom:idp_username"]
+        else:
+            audit_event_log["requestingUser"]["idpDisplayName"] = event["request"][
+                "userAttributes"
+            ]["custom:idp_display_name"]
```
[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/auth_function/lambda_function.py:95]: I see, I didn't notice this idp_user_id is in log already. Thanks for point that out.
Sure, then that seems to make sense with display_name; I assume it might be empty but does not matter.

File: server/auth_function/test/lamda_function_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/auth_function/test/lamda_function_test.py]: "True/False" is the default with one parameter. I see how it works now.

**Review by ianliuwk1019 — APPROVED** [review]

File: server/auth_function/lambda_function.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MCatherine1994) [reply server/auth_function/lambda_function.py:95]: I just saw your comment haha, our test bcsc account has display name that's why I feel it might be good to show it, it's a readable name.
